### DISCUSSION TITLE
bigskip doesn't work as environment

### DISCUSSION THREAD

**Discussion by pepa65** [sile-typesetter/sile#1600] [Q&A] [answered]
My sile document produces 1286 pages of pdf. When I use `\begin{bigskip}` after the `\begin{raggedright}` and `\end{bigskip}` before `\end{raggedright}`, it produces 1 empty page.

**Comment by alerque** [comment, marked as answer]
The `bigskip` command was not meant to be used as an environment — in other words it shouldn't be used to _wrap_ other content, it should be used _between_ other content. It outputs vertical space into the document where it was used then ends (discarding any content passed to it).

As such it would typically be used in the inline form `\bigskip` rather than with `\begin{...}\end{...}`. Because the SILE syntax is interchangeable you _can_ call it using the environment syntax, but you shouldn't put content inside it because it will just get ignored.

It looks like you might be trying to change either the leading (space between lines) or paragraph skip (space between paragraphs). These values can be adjusted with

    \set[parameter=document.baselineskip,value=3ex]

and

     \set[parameter=document.parskip,value=5ex]

respectively.

You should be able to set this inside a `raggedright` environment and the setting will end when you end the environment.

**Comment by pepa65** [comment]
I guess I can just do `\set[parameter=document.parskip, value=12pt]`

**Comment by alerque** [comment]
It might be worth noting that the space applied by `\bigskip` is actually flexible. The official definition is something like this:

```lua
SILE.registerCommand("bigskip", function (options, content)
    SILE.typesetter:leaveHmode()
    SILE.typesetter:pushExplicitVglue({ height = SILE.length.parse("12pt plus 4pt minus 4pt" ) })
end)
```

Note the allowance for the  space added to be stretched or shrunk by 4pt. If you want to replicate this in your inter-paragraph spacing you can use that syntax:

    \set[parameter=document.parskip, value=12pt plus 4pt minus 4pt]

Also note how the `\bigskip` command doesn't do anything with the `content` variable passed to it. If you wanted an environment that had a setting like this changed in it you could easily setup an environment of your own. For  example this would give  you something like you want:

```tex
\define[command=raggedloose]{\begin{raggedright}set[parameter=document.parskip, value=12pt plus 4pt minus 4pt]\process\end{raggedright}}
```

You could then use this new environment in place of ragged right + that parspace setting:

```tex
\begin{raggedloose}
this content will be ragged right

paragraphs will have ~12pt between them
\end{raggedloose}
```

**Comment by pepa65** [comment]
Thanks, excellent explanation and teaching.
I'm using 'plain' now, and have a mystifying phenomenon: the first line on each page is indented a bit less when preceded by a `\break`.

**Comment by alerque** [comment]
Are you using `\break` inline with other content to manually add line breaks or after paragraph breaks or times when the typesetter is in vertical mode (in which case it has the effect of a page-break)? Perhaps a small sample of your code and screen shot of the result showing the problem will make it easier to figure out what you're doing.

**Comment by pepa65** [comment]
At the moment I have this:
```
\begin{document}
\begin[right=true]{ragged}
\set[parameter=document.baselineskip,value=22pt]
\set[parameter=document.parskip,value=12pt]
\set[parameter=document.parindent,value=0]
\set[parameter=typesetter.widowpenalty,value=9000]
\set[parameter=typesetter.orphanpenalty,value=9000]
\font[family=Arundina Serif,language=und,script=Thai,size=14pt]

\define[command=gb]{\par\penalty[penalty=-500]}
\define[command=b]{\break}
\script[src=packages/counters]
\set-counter[id=ch,value=0]
\define[command=t]{\font[size=50pt,weight=900]{\process}\b}
\define[command=s]{\par\font[size=20pt,weight=900]{\process}\b}
\define[command=c]{\par\font[size=16pt,weight=900]{\process}\b}
\define[command=ch]{\increment-counter[id=ch]\c{\show-counter[id=ch] \process}}

\t{เพลงนมัสการ}
\c{ข้อสรรเสริญ}
สรรเสริญพระเจ้าผู้อำนวยพร\b
สรรเสริญพระเจ้าเหล่าประชากร\b
สรรเสริญองค์พระบิดาพระบุตร\b
สรรเสริญพระวิญญาณบริสุทธิ์\b
อาเมน
\s{เพลงเชิญชวนเข้านมัสการ 1-19}
\ch{เราจะยินดีและเราจะเปรมปรีดิ์}
เราจะยินดี และเราจะเปรมปรีดิ์\b
เราจะสดุดีสรรเสริญนามพระองค์\b
เพราะพระองค์ทรงนำ เราไปทุกวัน ทุกเวลา\b
เราจะยินดี และเรา จะเปรมปรีดิ์
\ch{ข้ายินดีเมื่อเขากล่าวแก่ข้าว่า}
[1] ข้ายินดี เมื่อเขากล่าวแก่ข้าว่า [x3]\b
ให้เราไป ยังนิเวศ พระเจ้าเถิด\b
[2] ร้องบทเพลง สาธุการแด่พระเจ้า [x3]\b
ให้เราไป ยังนิเวศ พระเจ้าเถิด
\ch{ฮาเลลูยาสรรเสริญ}
ฮาเลลูยาสรรเสริญ ฮาเลลูยาถวายสาธุการ\b
ขอถวายเกียรติแด่พระเจ้า ผู้ทรงอัศจรรย์\b
และบริสุทธิ์ ทรงประทานพระบุตรองค์เดียว\b
มาเป็นผู้ไถ่ ฮาเลลูยาสรรเสริญ ฮาเลลู\b
ถวายสาธุการ ฮาเลลูยาขอยกย่องพระนาม\b
พระสิริ คำสรรเสริญ แด่จอมราชา
\ch{มา นี่เป็นเวลานมัสการ}
มา นี่เป็นเวลานมัสการ\b
มา นี่เป็นเวลาถวายดวงใจ\b
มา นมัสการดั่งที่เป็น\b
มา ต่อหน้าพระองค์ดั่งที่เราเป็น มา
\gb
สักวันทุกลิ้นจะยอมรับว่าทรงเป็นพระเจ้า\b
สักวันทุกคนจะก้มกราบลง\b
แต่ทรัพย์สมบัติล้ำค่าคงเป็นของบรรดา\b
ผู้เลือกและรับพระองค์ [x2]
\gb
มา นี่เป็นเวลานมัสการ\b
มา นี่เป็นเวลาถวายดวงใจ\b
มา นมัสการดั่งที่เป็น\b
มา ต่อหน้าพระองค์ดั่งที่เราเป็น มา มา มา
```
And then much more before:
```
\end{ragged}
\end{document}
```
The `\gb` doesn't have the intended effect of course, because the breaks at the end of the verse lines have a lower penalty... What I'd like to achieve is no hanging indents, no page breaks within verses. I'm doing 2 versions, one where there is no page breaks within songs, and one where page breaks between verses are OK.

**Comment by simoncozens** [comment]
Thanks for this report - this is something I'd like to get working!

First problem: The reason you're getting a space at the beginning of the line is because of this:

```
Something\b<carriage return>
Something else
```

The carriage return is a whitespace; how to interpret end-of-line whitespace after a command is a perennial problem going back to issue #105, and I'm sorry about this. The workaround is to say:

```
Something\b%
Something else
```

Second problem: To avoid breaks within verses, try redefining `\b` as:

```
\define[command=b]{\break\novbreak}
```

**Comment by pepa65** [comment]
Thanks, that sounds like it could work. But I'm going to comment on #105 because I think the solution for that is really simple, and it would not require an empty comment...

**Comment by pepa65** [comment]
What I **really** need to solve the big problem (that the myriad of wkhtmltopdf based solutions can't seem to solve) is the page breaks at the desired places. If I'm going to use SILE I think I really need to understand the penalty system...

**Comment by pepa65** [comment]
Now if I run with the settings from the comment above but with \b being \break\novbreak I get no more "hanging indents". But somehow the text runs off the bottom of the pages, or way into the footer with the page number. The bigger problem is that I have no more verse breaks, every song is just 1 big verse...

Ideally there should be a template that works for people that want to make a song book, that is not an article or a regular book... Not having that, I wanted to start with 'plain' as then there are fewer defaults to contend with.

**Comment by Omikhleia** [comment]
Just passing by, but... What are the expectations in order for this issue to be closed? Initial question (corresponding to the issue title) was answered it seems. Other questions as well, on avoiding breaks, etc. I am not sure the last discussion is a new question.

**Comment by pepa65** [comment]
Postscript: using WeasyPrint was easier.
